demos/rag_agentic/notebooks/rag-agentic-mcp-demos.ipynb

```diff
@@ -0,0 +1,305 @@
+      "cell_type": "markdown",
+      "source": "### Query 2: (Agentic): `Search for the latest Red Hat OpenShift version on the Red Hat website. Summarize the version number and draft a short email to my team.`",
+      "metadata": {}
+    },
+    {
+      "id": "fe55883a-6887-43dd-9498-5333a51799e2",
+      "cell_type": "code",
+      "source": "# Code bellow written following examples here: https://llama-stack.readthedocs.io/en/latest/building_applications\nfrom llama_stack_client.lib.agents.agent import Agent\nfrom llama_stack_client.lib.agents.event_logger import EventLogger\nfrom llama_stack_client import LlamaStackClient\nimport argparse\nimport logging\nimport json\nimport os\nfrom dotenv import load_dotenv\n\nload_dotenv()\n\nlogger = logging.getLogger(__name__)\nlogger.setLevel(logging.INFO)\nstream_handler = logging.StreamHandler()\nstream_handler.setLevel(logging.INFO)\nformatter = logging.Formatter('%(message)s')\nstream_handler.setFormatter(formatter)\nlogger.addHandler(stream_handler)\n\n\nparser = argparse.ArgumentParser()\nparser.add_argument(\"-r\", \"--remote\", help=\"Uses the remote_url\", action=\"store_true\")\nparser.add_argument(\"-s\", \"--session-info-on-exit\", help=\"Prints agent session info on exit\", action=\"store_true\")\nparser.add_argument(\"-a\", \"--auto\", help=\"Automatically runs examples, and does not start a chat session\", action=\"store_true\")\nargs = parser.parse_args()\n\n# model=\"meta-llama/Llama-3.2-3B-Instruct\"\nmodel=\"ibm-granite/granite-3.2-8b-instruct\"\n\n# Connect to a llama stack server\nif args.remote:\n    base_url = os.getenv(\"REMOTE_BASE_URL\")\n    mcp_url = os.getenv(\"REMOTE_MCP_URL\")\nelse:\n    base_url=\"http://localhost:8321\"\n    mcp_url=\"http://host.containers.internal:8000/sse\"\n\nclient = LlamaStackClient(\n    base_url=base_url,\n    provider_data={\n        \"tavily_search_api_key\": os.getenv(\"TAVILY_API_KEY\")\n    })\nlogger.info(f\"Connected to Llama Stack server @ {base_url} \\n\")\n\nagent = Agent(\n    client=client,\n    model=model,\n    instructions = \"\"\"You are a helpful assistant. You have access to a number of tools.\n    Whenever a tool is called, be sure return the Response in a friendly and helpful tone.\n    When you are asked to search the web you must use a tool.\n    \"\"\" ,\n    tools=[\"builtin::websearch\"],\n    tool_config={\"tool_choice\":\"auto\"},\n    sampling_params={\"max_tokens\":4096}\n)\n\n\nsession_id = agent.create_session(session_name=\"Draft_email_with_latest_OCP_version\")\nprompt = \"\"\"Search for the web for the latest Red Hat OpenShift version on the Red Hat website. Summarize the version number and draft an email to convey this information.\"\"\"\nturn_response = agent.create_turn(\n    messages=[\n        {\n            \"role\":\"user\",\n            \"content\": prompt\n        }\n    ],\n    session_id=session_id,\n    stream=True,\n)\nfor log in EventLogger().log(turn_response):\n    log.print()\n",
```

Comment:
This code should be broken into multiple blocks explaining what each section of the code is doing.

Reply:
I have split the blocks out.